iio/iio.c

```diff
@@ -300,45 +307,52 @@ static struct iio_dev_priv *get_iio_device(struct iio_desc *desc,
 static int iio_read_all_attr(struct attr_fun_params *params,
 			     struct iio_attribute *attributes)
 {
-	/* TODO Not sure if working corectly */
-	return -EINVAL;
+	uint32_t written = 0;
 
-#if 0
-	int16_t i = 0, j = 0;
-	char local_buf[256];
-	int attr_length;
-	uint32_t *pattr_length;
+	/*
+	 * Not all devices have all types of attributes
+	 */
+	if (!attributes)
+		return -ENOENT;
 
-	while (attributes[i].name) {
-		attr_length = attributes[i].show(params->dev_instance,
-						 local_buf, params->len,
-						 params->ch_info,
-						 attributes[i].priv);
-		if (IS_ERR_VALUE(attr_length))
-			attr_length = snprintf(local_buf, params->len, "%d",
-					       attr_length);
-
-		attr_length += 1;//Add '\0' to the count
-		pattr_length = (uint32_t *)(params->buf + j);
-		if (j + 4 > params->len)
-			return -EINVAL;
-		*pattr_length = bswap_constant_32(attr_length);
-		j += 4;
-		if (attr_length >= 0) {
-			if (attr_length + j > params->len)
-				return -EINVAL;
-			sprintf(params->buf + j, "%s", local_buf);
-			if (attr_length & 0x3) /* multiple of 4 */
-				attr_length = ((attr_length >> 2) + 1) << 2;
-			j += attr_length;
+	for (struct iio_attribute *attr = attributes; attr->name; attr++) {
+		int attr_len;
+		uint32_t attr_len_be;
+		char *pattr_len;
+
+		// Allign to next multiple of 4
+		while (written & 0x3) {
+			params->buf[written++] = '\0';
+			if (written >= params->len)
+				return -ENOBUFS;
 		}
-		i++;
+
+		pattr_len = params->buf + written;
+
+		// reserve space for next attribute length
+		written += sizeof(attr_len_be);
+		if (written >= params->len)
+			return -ENOBUFS;
+
+		attr_len = attr->show(params->dev_instance,
+				      params->buf+written, params->len-written,
+				      params->ch_info, attr->priv);
+
+		if (attr_len < 0)
```

Comment:
The previous code printed the error code of show() into the buffer. To keep this functionality here we should check if return is != -ENOBUFS (which should be returned if out of buf space after/if iio_snprintf PR is accepted).